gdnative-sys/build.rs

```diff
@@ -76,304 +77,304 @@ mod header_binding {
 }
 
 mod api_wrapper {
-    #[derive(Debug, serde::Deserialize)]
-    struct ApiDescription<'a> {
-        #[serde(borrow)]
-        core: CoreApi<'a>,
-        #[serde(borrow)]
-        extensions: Vec<ExtensionApi<'a>>,
-    }
+    use lazy_static::lazy_static;
+    use proc_macro2::{Ident, TokenStream};
+    use quote::{format_ident, quote, ToTokens};
+    use regex::Regex;
+    use std::convert::AsRef;
+    use std::fs::File;
+    use std::io::Write as _;
+    use std::path;
 
     #[derive(Debug, serde::Deserialize)]
-    struct CoreApi<'a> {
-        #[serde(rename = "type")]
-        type_: &'a str,
-        version: Version,
-        #[serde(borrow)]
-        next: Option<Box<CoreApi<'a>>>,
-        #[serde(borrow)]
-        api: Vec<ApiFunction<'a>>,
-    }
-
-    impl<'a> CoreApi<'a> {
-        fn struct_type_name(&self, root: bool) -> String {
-            if root {
-                "godot_gdnative_core_api_struct".to_string()
-            } else {
-                format!(
-                    "godot_gdnative_core_{}_{}_api_struct",
-                    self.version.major, self.version.minor
-                )
-            }
-        }
+    struct ApiRoot {
+        core: Api,
+        extensions: Vec<Api>,
     }
 
     #[derive(Debug, serde::Deserialize)]
-    struct ExtensionApi<'a> {
-        name: Option<&'a str>,
+    struct Api {
+        name: Option<String>,
```

Comment:
Any reason these strings aren't borrowed like they were before? Are borrowed strings actually slower in this case?

Reply:
I felt having the `Api` own its data was more clear. All the `name` strings will be small, and this code generation only runs once at compile time, so I didn't give much consideration to performance.

Reply:
It's true that `gdnative-sys` is not the biggest offender when it comes to build times, but such a change still seems rather needless. I can't really see why a `String` is in any way more clear than a `&'_ str`.

Reply:
The code currently reads the entire JSON file into memory, and then runs serde_json on that memory, and serde_json parses out the various API structs, which hold references back to the full JSON file still in memory.

The proposed code in this PR parses from a file reader directly into the API structs, and thus the API structs cannot hold references, they must hold the actual data from the file, because the entire file is no longer held in memory. This _probably_ (I haven't measured) results in less memory usage, and I don't know how it would effect the speed.

Reply:
I see.